src/utils/TodoContext.tsx

```diff
@@ -0,0 +1,100 @@
+import React, {
+  createContext,
+  useContext,
+  useState,
+  FC,
+  useEffect,
+} from 'react';
+import { Todo } from '../types/Todo';
+import { TodosContextType } from '../types/TodosContextType';
+import { Filter } from '../types/Filter';
+import { getTodos } from '../api/todos';
+
+const initialTodos: Todo[] = [];
+
+const TodosContext = createContext<TodosContextType>({
+  todos: initialTodos,
+  filter: Filter.ALL,
+  addTodo: () => {},
+  removeTodo: () => {},
+  setTodos: () => {},
+  setFilter: () => {},
+  query: '',
+  setQuery: () => {},
+  error: null,
+  setError: () => {},
+  isLoading: false,
+  setIsLoading: () => {},
+});
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const TodosProvider: FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>(() => {
+    const storedTodos = localStorage.getItem('todos');
+
+    return storedTodos ? JSON.parse(storedTodos) : [];
+  });
+  const [filter, setFilter] = useState<Filter>(Filter.ALL);
+  const [query, setQuery] = useState('');
+  const [error, setError] = useState<{ message: string } | null>(null);
+  const [isLoading, setIsLoading] = React.useState(false);
+
+  const addTodo = (newTodo: Todo) => {
+    setTodos(prevTodos => [...prevTodos, newTodo]);
+  };
+
+  const removeTodo = (id: number) => {
+    setTodos(todos.filter(todo => todo.id !== id));
```

Comment:
Use functional setState when you need to set state depends on the previous one

Fix it in all places